### DISCUSSION TITLE
Borrower coin selects before deciding on the fee

### DISCUSSION THREAD

**Discussion by luckysori** [comit-network/baru#57] [Ideas]
Very early on in the design of the loan protocol we decided to let the borrower control the transaction fee, since they are the party contributing the bitcoin to the loan transaction (Liquid transaction fees can only be paid with bitcoin).

Specifically, the borrower gets to choose the amount of satoshis per vbyte for the fee, and once the loan transaction is constructed that value is used to calculate the actual fee.

The borrower also selects some inputs to pay for the collateral they put up, but they do so _before_ the actual fee is calculated. This can be problematic if they forget to account for this yet-to-be-calculated fee during coin selection. Requiring users of the library to understand how the protocol works and that they must provide inputs covering the collateral amount + an undefined fee is not good UX.

I see a few alternatives to solve this problem (if we want to solve it), but there could be more:

1. Let the borrower commit to a fee earlier on, so that they can account for the concrete value when coin selecting. Calculating a reasonable fee would be the borrower's responsibility, but without knowing the exact structure of the transaction, which is something that we currently take advantage of here: https://github.com/comit-network/baru/blob/4bc7525bbef5255eefc7c23c668701fdf9de0ae1/src/loan.rs#L1104-L1106
2. Let the lender choose and pay for the fee instead. It makes constructing the transaction a bit more complicated (hopefully simple again once we use PSETs), but the lender knows the structure of the transaction by the time they add their inputs.
3. Do not add a fee output directly to the loan transaction and instead use a fee bumping technique such as Child Pays for Parent ([CPFP](https://bitcoinops.org/en/topics/cpfp/)). We've talked about something like this in the past, but we've never tried it AFAIK.

**Comment by thomaseizinger** [comment]
> Let the lender choose and pay for the fee instead. It makes constructing the transaction a bit more complicated (hopefully simple again once we use PSETs), but the lender knows the structure of the transaction by the time they add their inputs.

I think I am in favor of this. You could argue that the transaction fees are "operating costs" and need to be covered by the interest that the lender is earning.

Why would it make constructing the transaction more complicated?

**Reply by luckysori** [reply]
> Why would it make constructing the transaction more complicated?

If the lender pays for the fee, then they need to add a bitcoin input on top of their principal input, plus a likely change output for the leftovers after covering the fee.

**Reply by luckysori** [reply]
Also, do you think doing `3` is a bad idea or the wrong approach? The lender would still be responsible for paying for the fees since they're the one who broadcast the loan transaction.
